src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs

```diff
@@ -273,18 +335,917 @@ private static IEnumerable<SyntaxNode> GetOwningNodes(SyntaxNode root, int posit
+            }
+
+            if (selectedNode is IfStatementSyntax ifStatementNode)
+            {
+                // If this is just an if without else
+                // e.g.
+                // before:
+                // if$$ (true)
+                //   Print("Hello");
+                // after:
+                // if (true)
+                // {
+                //      $$
+                // }
+                //   Print("Hello");
+                if (ifStatementNode.Else == null)
+                {
+                    return ReplaceStatementOwnerAndInsertStatement(document,
+                        root,
+                        selectedNode,
+                        WithBracesForEmbeddedStatementOwner(selectedNode),
+                        ImmutableArray<SyntaxNode>.Empty.Add(statement),
+                        cancellationToken);
+                }
+
+                // If this IfStatement has else statement after
+                // e.g.
+                // before:
+                // if (true)
+                //     print("Hello");
+                // else {}
+                // after:
+                // if (true)
+                // {
+                //     $$
+                //     print("Hello");
+                // }
+                // else {}
+                var newRoot = ReplaceNodeAndFormat(
+                    document,
+                    root,
+                    selectedNode,
+                    WithBracesForEmbeddedStatementOwner(selectedNode, statement),
+                    cancellationToken);
+                var nextCaretPosition = GetOpenBraceSpanEnd(newRoot);
+                return (newRoot, nextCaretPosition);
+            }
+
+            if (selectedNode is ElseClauseSyntax elseClauseNode)
+            {
+                // If this is an 'els$$e if(true)' statement,
+                // then treat it as the selected node is the nested if statement
+                if (elseClauseNode.Statement is IfStatementSyntax)
+                {
+                    return AddBraceToEmbeddedStatementOwner(document, root, elseClauseNode.Statement, cancellationToken);
+                }
+
+                // Otherwise, it is just an ending else clause
+                // e.g. before:
+                // if (true)
+                // {
+                // } els$$e
+                // var i = 10;
+                // after:
+                // if (true)
+                // {
+                // } els$$e
+                // {
+                //      $$
+                // }
+                // var i = 10;
+                return ReplaceStatementOwnerAndInsertStatement(document,
+                    root,
+                    selectedNode,
+                    WithBraces(selectedNode),
+                    ImmutableArray<SyntaxNode>.Empty.Add(statement),
+                    cancellationToken);
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
```

Comment:
1. this method is great in terms of explaining what it is doing.
2. one recommendatino is that each case is rather large.  consider extracting helper methods for each so the dispatch methods are short and clear, and the helper methods are single purpose.